[FLINK-5972] Don't allow shrinking merging windows

aljoscha opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

aljoscha commented:
@StephanEwen this is the better alternative to #3535 in that it doesn't silently drop windows and throws an exception when window shrinking occurs.

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+ * <p><b>Important:</b>This test must always be kept up-to-date with
+ * {@link WindowOperatorContractTest}.
+ */
+public class EvictingWindowOperatorContractTest extends TestLogger {
```
nitpick: The line spacing between the tests / methods in this class is also generally a bit off, occasionally there's 2-line spacings.

tzulitai replied:
Also, should we consider moving the static mock / matcher util methods to a separate class, similar to the `StreamRecordMatchers` class?

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+import static org.mockito.Matchers.anyLong;
+import static org.mockito.Mockito.*;
+
+import com.google.common.collect.Lists;
```
I think there was some sort of agreement in avoiding Guava in Flink.

aljoscha replied:
I'm replacing it by `new ArrayList<>(Arrays.asList(...))`

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+		verify(mockAssigner, times(1)).assignWindows(eq(0), eq(10L), anyAssignerContext());
+
+		assertThat(testHarness.getSideOutput(lateOutputTag),
+				containsInAnyOrder(isStreamRecord(0, 5L), isStreamRecord(0, 10L)));
```
Isn't the ordering of side outputs supposed to be deterministic? Why do we need to use `containsInAnyOrder`?

aljoscha replied:
Fixing

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+		testHarness.processElement(new StreamRecord<>(0, 0L));
+
+		verify(mockAssigner).mergeWindows(eq(Lists.newArrayList(new TimeWindow(2, 4))), anyMergeCallback());
+		verify(mockAssigner).mergeWindows(eq(Lists.newArrayList(new TimeWindow(2, 4), new TimeWindow(0, 2))), anyMergeCallback());
```
Avoid Guava in Flink. `Arrays.asList(...)` should do fine.

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+		shouldMergeWindows(
+				mockAssigner,
+				Lists.newArrayList(new TimeWindow(0, 2), new TimeWindow(2, 4)),
+				Lists.newArrayList(new TimeWindow(0, 2), new TimeWindow(2, 4)),
```
Same here, avoid Guava ;-)

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+			public TriggerResult answer(InvocationOnMock invocation) throws Exception {
+				return TriggerResult.FIRE;
+			}
+		}).when(mockTrigger).onElement(Matchers.<Integer>anyObject(), anyLong(), anyTimeWindow(), anyTriggerContext());
```
Why not reuse `shouldFireOnElement` here?

aljoscha replied:
fixing

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+	}
+
+	private static <T> void shouldContinueOnElement(Trigger<T, TimeWindow> mockTrigger) throws Exception {
+		when(mockTrigger.onElement(Matchers.<T>anyObject(), anyLong(), Matchers.<TimeWindow>any(), anyTriggerContext())).thenReturn(TriggerResult.CONTINUE);
```
can we resuse `anyTimeWindow()` to replace `Matchers.<TimeWindow>any()`, just because we've defined the method in this class anyways :-)?

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+	}
+
+
+	private  void testEmittingFromWindowFunction(TimeDomainAdaptor timeAdaptor) throws Exception {
```
redundant space between `private` and `void`

aljoscha replied:
fixing

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+
+	/**
+	 * Verify that we neither invoke the trigger nor the window function if a timer
+	 * for an empty merging window.
```
seems like a missing word here: "if a timer _fires_ for an empty merging window"

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+	@Test
+	public void testRejectShrinkingMergingEventTimeWindows() throws Exception {
+		testRejectShrinkingMergingWindows(new EventTimeAdaptor());
+
```
This can be private

tzulitai replied:
This can be private.

tzulitai replied:
private :) All of the testXXXXXXX branch methods for testing below can be private, I think.

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+		testDeletedTimerDoesNotFire(new ProcessingTimeAdaptor());
+	}
+
+	public void testDeletedTimerDoesNotFire(TimeDomainAdaptor timeAdaptor) throws Exception {
```
This can be private.

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+	/**
+	 * Verify that windows are merged eagerly, if possible.
+	 */
+	public void testWindowsAreMergedEagerly(final TimeDomainAdaptor timeAdaptor) throws Exception {
```
Private :)

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorContractTest.java:
```diff
@@ -1577,6 +1585,91 @@ public TriggerResult answer(InvocationOnMock invocation) throws Exception {
+	 * A misbehaving {@code WindowAssigner} can cause a window to become late by merging if
+	 * it moves the end-of-window time before the watermark. This verifies that we don't allow that.
+	 */
+	public void testRejectShrinkingMergingWindows(final TimeDomainAdaptor timeAdaptor) throws Exception {
```
This can be private.

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorContractTest.java:
```diff
@@ -1577,6 +1585,91 @@ public TriggerResult answer(InvocationOnMock invocation) throws Exception {
+				context.getPartitionedState(valueStateDescriptor).update("hello");
+				return TriggerResult.CONTINUE;
+			}
+		}).when(mockTrigger).onElement(Matchers.<Integer>anyObject(), anyLong(), anyTimeWindow(), anyTriggerContext());
```
I wonder if we want to have a common class to contain all these static mock util methods for these kind of bigger stubs too.
Together with `EvictingWindowOperatorContractTest` that could save quite a bit of LoC and easier to read.

tzulitai commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorContractTest.java:
```diff
@@ -1577,6 +1585,91 @@ public TriggerResult answer(InvocationOnMock invocation) throws Exception {
+
+		shouldMergeWindows(
+				mockAssigner,
+				Lists.newArrayList(new TimeWindow(8, 11), new TimeWindow(10, 13)),
```
Avoid Guava.

aljoscha commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperator.java:
```diff
@@ -122,6 +121,20 @@ public void processElement(StreamRecord<IN> element) throws Exception {
+											" window: " + mergeResult);
+								} else if (!windowAssigner.isEventTime() && mergeResult.maxTimestamp() <= internalTimerService.currentProcessingTime()) {
+									throw new UnsupportedOperationException("The end timestamp of a " +
+											"processing-time window cannot become earlier than the current procesing time" +
```
Fixing

aljoscha commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperator.java:
```diff
@@ -122,6 +121,20 @@ public void processElement(StreamRecord<IN> element) throws Exception {
+											"processing-time window cannot become earlier than the current procesing time" +
+											"by merging. Current processing time: " + internalTimerService.currentProcessingTime() +
+											" window: " + mergeResult);
+
```
Fixing

aljoscha commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java:
```diff
@@ -138,7 +138,7 @@
 	 * be emitted to this.
 	 */
-	private final OutputTag<IN> lateDataOutputTag;
+	protected final OutputTag<IN> lateDataOutputTag;
```
It is protected so that `EvictingWindowOperator` can access it. I could make it package private, though.

tzulitai replied:
Ah, sorry I mistakened that it was exposed for testing. `protected` would be fine then.

aljoscha commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java:
```diff
@@ -352,6 +352,20 @@ public void processElement(StreamRecord<IN> element) throws Exception {
+									" window: " + mergeResult);
+						} else if (!windowAssigner.isEventTime() && mergeResult.maxTimestamp() <= internalTimerService.currentProcessingTime()) {
+							throw new UnsupportedOperationException("The end timestamp of a " +
+									"processing-time window cannot become earlier than the current procesing time" +
```
Fixing

aljoscha commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java:
```diff
@@ -574,7 +588,7 @@ private void emitWindowContents(W window, ACC contents) throws Exception {
 	 * @param element skipped late arriving element to side output
 	 */
-	private void sideOutput(StreamRecord<IN> element){
+	protected void sideOutput(StreamRecord<IN> element){
```
Same as for `lateDataOutputTag`. It's used by the `EvictingWindowOperator`. I can also make it package private.

aljoscha commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.junit.Assert.*;
+import static org.mockito.Matchers.anyLong;
+import static org.mockito.Mockito.*;
```
Done. I would like to have strict checkstyle for tests as well ... 😅

aljoscha commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java:
```diff
@@ -0,0 +1,2654 @@
+	}
+
+	/**
+	 * Verify that there is no late-date side output if the {@code WindowAssigner} does
```
Fixing

aljoscha commented:
Thanks for the thorough review, @tzulitai!

And shame over my head for all the code duplication 😱 

I now refactored the contract tests into a base class and addressed all the issues that you found (hopefully).

The failing `WindowOperatorTest` tests actually discovered a real problem, I wasn't taking the allowed lateness into account when calculating if a merged window would be late. I modified the new tests in in the contract test to verify that we now have the correct behaviour.

Could you maybe have another look?

tzulitai commented:
Thanks a lot for addressing my comments @aljoscha!

That's so much cleaner now :-D Really like the refactoring in the fixes.
I actually didn't realize the mock stubs and util methods were already in `WindowOperatorContractTest` just looking at the previous diffs :P

I did one final pass on the changes and didn't find anything. +1 to merge after Travis turns green.

tzulitai commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperator.java:
```diff
@@ -122,6 +121,19 @@ public void processElement(StreamRecord<IN> element) throws Exception {
 									Collection<W> mergedWindows, W stateWindowResult,
 									Collection<W> mergedStateWindows) throws Exception {
+
+								if ((windowAssigner.isEventTime() && mergeResult.maxTimestamp() + allowedLateness <= internalTimerService.currentWatermark())) {
```
Ah, I see. Good seeing the tests at work catching this :-D

tzulitai commented:
Thanks a lot for the extra pair of eyes on this @kl0u!
I can fix the typo and proceed to merge this to `master` and `release-1-2`.

tzulitai commented:
Merging ..

aljoscha commented:
Thanks for merging, @tzulitai! 😃

kl0u commented:
Thanks @tzulitai

tzulitai commented:
FYI: I managed to backport the changes (including the contract test refactorings) back to `release-1.2`.
But since it involved quite a few conflict resolvings, I'm waiting on a local Travis run just to be sure before merging the backport: https://travis-ci.org/tzulitai/flink/builds/214315026
